README.md
cerus-server


changelog.md
# Changelog

## v0.1.1 (23 Jan. 2018)
- (New) Redid the whole module again.
- (New) Added eslint and fixed the issues it found.
- (Enh) Removed a unused function.
- (Enh) Moved the address class.
- (Enh) Removed some unused plugin dependencies.

## v0.1.0 (1 Jan. 2018)
- (New) Redid the whole module.
- (Enh) Updated some dev dependencies.
- (Fix) Fixed some tests.

## v0.0.3 (17 Nov. 2017)
- (New) Added tests.
- (New) Added the functions #event(), #maxheaders(), #maxconnections(), #alive() and #timeout().
- (Enh) #start() will now test if the server is listening and it's arguments.
- (Enh) #start() can now be supplied with a custom port.
- (Enh) #stop(), #connections() and #address will now test if the server is listening.

## v0.0.2 (2 Nov. 2017)
- (Fix) Fixed the package.json.
- (Fix) Fixed an issue with requiring the wrong file.

## v0.0.1 (2 Nov. 2017)
- (New) Created the module.
- (New) Added the package.json.
- (New) Added the promise class.
- (New) Added the .gitignore.


tests/server.test.js
var expect = require("chai").expect;
var http = require("http");
var cerus = require("cerus")();
cerus.use(require("cerus-request")());
var server = require("../index")();
server._init(cerus);

describe("server", function() {
	describe("constructor", function() {
		context("with no parameters", function() {
			it("shouldn't throw an error", function() {
				var func = function() {
					server._init(cerus);
					server.server();
				}

				expect(func).to.not.throw();
			});
		});
	});

	describe("#start", function() {
		context("with no parameters", function() {
			it("shouldn't throw an error", function(done) {
				var func = function() {
					server._init(cerus);
					server.server().start()
					.then(function() {
						server.server().stop();
						done();
					});
				}

				expect(func).to.not.throw();
			});
		});

		context("with a server running and no parameters", function() {
			it("should throw an error", function(done) {
				server._init(cerus);
				server.server().start()
				.then(function() {
					var func = function() {
						server.server().start();
					}

					expect(func).to.throw();
					done();
					server.server().stop();
				});
			});
		});

		context("with a number as parameter", function() {
			it("shouldn't throw an error", function(done) {
				var func = function() {
					server._init(cerus);
					server.server().start(12345)
					.then(function() {
						server.server().stop();
						done();
					});
				}

				expect(func).to.not.throw();
			});
		});

		context("with a string as parameter", function() {
			it("should throw a TypeError", function() {
				var func = function() {
					server._init(cerus);
					server.server().start("error");
				}

				expect(func).to.throw();
			});
		});
	});

	describe("#callback", function() {
		context("with no parameters", function() {
			it("should throw a TypeError", function() {
				var func = function() {
					server._init(cerus);
					server.server().callback();
				}

				expect(func).to.throw();
			});
		});

		context("with a non-function as parameter", function() {
			it("should throw a TypeError", function() {
				var func = function() {
					server._init(cerus);
					server.server().callback("error");
				}

				expect(func).to.throw();
			});
		});

		context("with no server running and a function as parameter", function() {
			it("should not be called", function() {
				server._init(cerus);
				server.server().callback(function() {
					throw new Error("this should not be called");
				});
			});
		});

		context("with a server running and a function as parameter", function() {
			it("should be called", function(done) {
				server._init(cerus);
				server.server().callback(function(req, res) {
					res.end();
					done();
				});
				server.server().start()
				.then(function() {
					cerus.request()
					.send(function() {
						server.server().stop();
					});
				});
			});
		});
	});

	describe("#stop", function() {
		context("with no running server and no parameters", function() {
			it("should throw an error", function() {
				var func = function() {
					server._init(cerus);
					server.server().stop();
				}

				expect(func).to.throw();
			});
		});

		context("with a running server and no parameters", function() {
			it("shouldn't throw an error", function(done) {
				var func = function() {
					server._init(cerus);
					server.server().start()
					.then(function() {
						server.server().stop()
						.then(function() {
							done();
						});
					});
				}

				expect(func).to.not.throw();
			});
		});
	});

	describe("#listening", function() {
		context("with no running server and no parameters", function() {
			it("should return false", function() {
				server._init(cerus);
				var listening = server.server().listening();
				expect(listening).to.equal(false);
			});
		});

		context("with a running server and no parameters", function(done) {
			it("should return true", function() {
				server._init(cerus);
				server.server().start()
				.then(function() {
					var listening = server.server().listening();
					expect(listening).to.equal(true);
					server.server().stop();
					done();
				});
			});
		});
	});

	describe("#maxheaders", function() {
		context("with no parameters", function() {
			it("should return the default value", function() {
				server._init(cerus);
				var max = server.server().maxheaders();
				expect(max).to.equal(null);
			});
		});

		context("with 2500 as parameter", function() {
			it("should return the new value", function() {
				server._init(cerus);
				var max = server.server().maxheaders(2500);
				expect(max).to.equal(2500);
			});
		});
	});

	describe("#maxconnections", function() {
		context("with no parameters", function() {
			it("should return the default value", function() {
				server._init(cerus);
				var max = server.server().maxconnections();
				expect(max).to.equal(undefined);
			});
		});

		context("with 100 as parameter", function() {
			it("should return the new value", function() {
				server._init(cerus);
				var max = server.server().maxconnections(100);
				expect(max).to.equal(100);
			});
		});
	});

	describe("#alive", function() {
		context("with no parameters", function() {
			it("should return the default value", function() {
				server._init(cerus);
				var max = server.server().alive();
				expect(max).to.equal(5000);
			});
		});

		context("with 8000 as parameter", function() {
			it("should return the new value", function() {
				server._init(cerus);
				var max = server.server().alive(8000);
				expect(max).to.equal(8000);
			});
		});
	});

	describe("#timeout", function() {
		context("with no parameters", function() {
			it("should return the default value", function() {
				server._init(cerus);
				var max = server.server().timeout();
				expect(max).to.equal(120000);
			});
		});

		context("with 150000 as parameter", function() {
			it("should return the new value", function() {
				server._init(cerus);
				var max = server.server().timeout(150000);
				expect(max).to.equal(150000);
			});
		});
	});

	describe("#connections", function() {
		context("with no running server", function() {
			it("should throw an error", function() {
				var func = function() {
					server._init(cerus);
					server.server().connections();
				}

				expect(func).to.throw();
			});
		});

		context("with a running server and 0 connections", function() {
			it("should return the number 0", function(done) {
				server._init(cerus);
				server.server().start()
				.then(function() {
					server.server().connections()
					.then(function(value) {
						expect(value).to.equal(0);
						server.server().stop();
						done();
					});
				});
			});
		});
	});

	describe("#address", function() {
		describe("constructor", function() {
			context("with no running server", function() {
				it("should throw an error", function() {
					var func = function() {
						server._init(cerus);
						server.server().address();
					}

					expect(func).to.throw();
				});
			});
		});

		describe("#address", function() {
			context("with no running server", function() {
				it("should throw an error", function() {
					var func = function() {
						server._init(cerus);
						server.server().address().address();
					}

					expect(func).to.throw();
				});
			});

			context("with a running server", function() {
				it("should return an object with the string object.address", function(done) {
					server._init(cerus);
					server.server().start()
					.then(function() {
						var address = server.server().address().address();
						expect(address).to.be.a("string");
						server.server().stop();
						done();
					});
				});
			});
		});

		describe("#port", function() {
			context("with no running server", function() {
				it("should throw an error", function() {
					var func = function() {
						server._init(cerus);
						server.server().address().port();
					}

					expect(func).to.throw();
				});
			});

			context("with no running server", function() {
				it("should return an object with the number object.port", function(done) {
					server._init(cerus);
					server.server().start()
					.then(function() {
						var port = server.server().address().port();
						expect(port).to.be.a("number");
						server.server().stop();
						done();
					});
				});
			});
		});

		describe("#family", function() {
			context("with no running server", function() {
				it("should throw an error", function() {
					var func = function() {
						server._init(cerus);
						server.server().address().family();
					}

					expect(func).to.throw();
				});
			});

			context("with no running server", function() {
				it("should return an object with the string object.family", function(done) {
					server._init(cerus);
					server.server().start()
					.then(function() {
						var family = server.server().address().family();
						expect(family).to.be.a("string");
						server.server().stop();
						done();
					});
				});
			});
		});
	});
});

lib/server.js
var http = require("http");

/**
 * This is the server class. With you can easily create and manage a http server. This class is 
 * also needed when working with the cerus-router. When there is a request the callback function is
 * called. This callback has to be set if you want to receive incoming requets. It must be noted 
 * that this name can be confusing and therefor a different module is on it's way. This means this 
 * class will be deprecated somewhere in the near future.
 * @example
 * // in this example a server is created and started
 * cerus.server().callback(function(req, res) {});
 * // -> now all incoming request will call this function
 * 
 * cerus.server().start();
 * // -> starts the server on port 80
 * @class server
 */
class server {
	constructor(cerus) {
		this._server = http.createServer(function(req, res) {
			if(typeof this._func === "function") {
				this._func(req, res);
			}
		}.bind(this));
		this._port = 80;
		this._cerus = cerus;
	}

	/**
	 * This function sets the callback function for the server. This function will be called when 
	 * there is a new request. The function will be called with to parameters: req and res. The 
	 * req parameter contains the request (http.IncomingMessage) and the res parameter the response
	 * (http.ServerResponse). For a custom and better req and res take a look at the cerus-router 
	 * module. 
	 * @example
	 * cerus.server().callback(function(req, res) {});
	 * // -> this will set the callback function
	 * @summary Sets the callback that will be called for incoming requests.
	 * @param {Function} func The callback function.
	 * @function callback
	 */
	callback(func) {
		if(typeof func !== "function") {
			throw new TypeError("the argument func must be a function");
		}

		this._func = func;
	}

	/**
	 * This function is used to start the server. You can also supply a port if you don't want to 
	 * use the one that is currently set. The server cannot be started when it is already running.
	 * This function will also return a promise. This promise will call the "started" event when 
	 * the server has been started.
	 * @example
	 * cerus.server().start(); // or cerus.server().listen();
	 * // -> starts the server on the default port 80
	 * @summary Starts the server.
	 * @alias listen
	 * @param {Number} (port) The port that the server will use.
	 * @return {Promise} This function will return a promise.
	 * @function start
	 */
	start(port) {
		if(this.listening()) {
			throw new Error("the server was already listening");
		}

		if(port !== undefined && typeof port !== "number") {
			throw new TypeError("the argument port must be a number");
		}

		return this._cerus.promise(function(event) {
			this._server.listen(port || this._port, function() {
				event("started");
			});
		}.bind(this));
	}

	listen(port) {
		return this.start(port);
	}

	/**
	 * This function will stop the server. It can only be called if the server is already running. 
	 * It'll also return a promise that calls the "stopped" event when the function has finished 
	 * stopping the server.
	 * @example
	 * // with a server already running
	 * cerus.server().stop(); // or cerus.server().end();
	 * // -> stops the server
	 * @summary Stops the server.
	 * @alias end
	 * @return {Promise} This function will return a promise.
	 * @function stop
	 */
	stop() {
		if(!this.listening()) {
			throw new Error("the server never started listening");
		}

		return this._cerus.promise(function(event) {
			this._server.close(function() {
				event("stopped");
			});
		}.bind(this));
	}

	end() {
		return this.stop();
	}

	/**
	 * This is a getter and setter for the port that will be used for the server. It is possible to
	 * not use this port by supplying a port to the .start() function directly.
	 * @summary The getter/setter for the server port.
	 * @param {Number} (port) The new server port.
	 * @return {Number} The server port.
	 * @function port
	 */
	port(port) {
		if(typeof port === "number") {
			this._port = port;
		}

		return this._port;
	}

	/**
	 * With this function you can listen for server events. It'll return a promise that is called 
	 * on a number of events. The "error" event is called when there was an error in the server. It
	 * is called with the error as parameter. The "close" event is called when the server was 
	 * closed. The "listening" event is called when the server has started listening. The 
	 * "connection" event is called when there was a new connection together with the new socket as
	 * paramater.
	 * @summary Returns a promise that is called on every event.
	 * @return {Promise} This function will return a promise.
	 * @function event
	 */
	event() {
		return this._cerus.promise(function(event) {
			this._server.on("error", function(error) {
				event("error", error);
			});

			this._server.on("close", function() {
				event("close");
			});

			this._server.on("listening", function() {
				event("listening");
			});

			this._server.on("connection", function(socket) {
				event("connection", socket);
			});
		}.bind(this));
	}

	/**
	 * This function returns if the server is currently listening for requests. This means that 
	 * this function returns if the server has started or if it's currently stopped.
	 * @summary Returns if the server is listening.
	 * @return {Boolean} If the server is listening.
	 * @function listening
	 */
	listening() {
		return this._server.listening;
	}

	/**
	 * This is a getter and setter for the maximum amount of headers that the server will accept in 
	 * a HTTP request. The default maximum amount of headers is 2000.
	 * @summary The getter/setter for the maximum amount of headers.
	 * @param {Number} (maxheaders) The new maximum amount of headers.
	 * @return {Number} The maximum amount of headers.
	 * @function maxheaders
	 */
	maxheaders(maxheaders) {
		if(typeof maxheaders === "number") {
			this._server.maxHeadersCount = maxheaders;
		}

		return this._server.maxHeadersCount;
	}

	/**
	 * This is a getter and setter for the maximum amount of connections that the server will accept 
	 * before refusing requests. By default there is no maximum amount of connections.
	 * @summary The getter/setter for the maximum amount of connections.
	 * @param {Number} (maxconnections) The new maximum amount of connections.
	 * @return {Number} The maximum amount of connections.
	 * @function maxconnections
	 */
	maxconnections(maxconnections) {
		if(typeof maxconnections === "number") {
			this._server.maxConnections = maxconnections;
		}

		return this._server.maxConnections;
	}

	/**
	 * This function will return the current amount of connections. The connections will be 
	 * returned using a promise. The amount of connections will be a parameter in the connections 
	 * event. The amount of connections can only be fetched if the server is listening.
	 * @summary Returns the amount of active connections.
	 * @return {Number} The amount of connections.
	 * @function connections
	 */
	connections() {
		if(!this.listening()) {
			throw new Error("the server hasn't started listening");
		}
		
		return this._cerus.promise(function(event) {
			this._server.getConnections(function(err, count) {
				
				if(err) {
					event("error");
				}
				else {
					event("connections", count);
				}
			});
		}.bind(this));
	}

	/**
	 * This is a getter and setter for the number of milliseconds of inactivity the server needs to 
	 * wait for incoming data. If it takes more than the alive time the request is presumed 
	 * inactive. The default time is 5000 milliseconds (5 seconds).
	 * @summary The getter/setter for the keep alive timeout time.
	 * @param {Number} (alive) The keep alive timeout time.
	 * @return {Number} The keep alive timeout time.
	 * @function alive
	 */
	alive(alive) {
		if(typeof alive === "number") {
			this._server.keepAliveTimeout = alive;
		}

		return this._server.keepAliveTimeout;
	}

	/**
	 * This is a getter and setter for the number of milliseconds of inactivity a socket can use 
	 * before being presumed to have timed out. When a socket is presumed to have timed out the 
	 * connection is destroyed. The default time 120000 (2 minutes).
	 * @summary The getter/setter for the timeout time.
	 * @param {Number} (timeout) The timeout time.
	 * @return {Number} The timeout time.
	 * @function timeout
	 */
	timeout(timeout) {
		if(typeof timeout === "number") {
			this._server.setTimeout(timeout);
		}

		return this._server.timeout;
	}

	/**
	 * This function will return the address class for this server. The address class contains 
	 * information about the address the server is using. This class can only be returned of the 
	 * server is listening.
	 * @summary Returns the server.address class.
	 * @return {Class} The server.address class.
	 * @function address
	 */
	address() {
		if(!this.listening()) {
			throw new Error("the server never started listening");
		}

		if(this._address === undefined) {
			this._address = new address(this._server);
		}

		return this._address;
	}
}

module.exports = server;

/**
 * This is the address class for the server. It contains information about the server, like the 
 * port and ip address.
 * @class server.address
 */
class address {
	constructor(server) {
		this._address = server.address();
	}

	/**
	 * This function will return the port the server is currently listening on.
	 * @summary Returns the port the server is using.
	 * @return {Number} The port the server is using.
	 * @function port
	 */
	port() {
		return this._address.port;
	}

	/**
	 * This function will return if the ip the server is currently using is IPv4 or IPv6.
	 * @summary Returns the family of the server's ip.
	 * @return {String} The family of the server's ip.
	 * @function family
	 */
	family() {
		return this._address.family;
	}

	/**
	 * This function will return the ip the server is currently using.
	 * @summary Returns the ip of the server.
	 * @return {String} The ip of the server.
	 * @function address
	 */
	address() {
		return this._address.address;
	}
}
